### DISCUSSION TITLE
Adding Proper Types to MockInstance When Using `jest.fn`

### DISCUSSION THREAD

**Discussion by pcbowers** [help-me-mom/ng-mocks#9318] [Q&A] [answered]
Currently, when I use `MockInstance` with `jest.fn`:

```ts
MockInstance(UserService, 'getUser', jest.fn())
```

The resulting type of this function is:

```ts
jest.Mock<any, any, any>
```

Ideally, it would actually be mocked to be:

```ts
jest.MockedFn<() => Observable<User>>
```

The best way I've found to get it to be typed properly is to add this somewhere in a typescript file, probably some global `.d.ts` file:

```ts
declare module 'ng-mocks' {
  function MockInstance<T extends object, K extends keyof T, S extends T[K]>(
    instance: AnyType<T>,
    name: K,
    stub: jest.Mock
  ): jest.MaybeMocked<S>;
}
```

That said, I was curious if anyone had a better solution out there to ensure it was typed properly, or if maybe I was missing something. Is there a better way to do this? If not, is this an overload that could be added to `ng-mocks`, or, since it's specific to jest, is that not something that is done?

**Comment by satanTime** [comment, marked as answer]
A Jest-specific overload in ng-mocks is not necessary. The current `MockInstance` overload preserves the exact type of the stub passed as its third argument. The broad type comes from calling `jest.fn()` without type parameters.

With the current Jest typings, type the return value and parameters when creating the mock:

```ts
type GetUser = UserService['getUser'];

const getUser = jest.fn<
  ReturnType<GetUser>,
  Parameters<GetUser>
>();

const mockGetUser = MockInstance(
  UserService,
  'getUser',
  getUser,
);

mockGetUser.mockReturnValue(of(user));
```

`mockGetUser` keeps the typed return value and argument list, while `MockInstance` still validates that it is assignable to `UserService.getUser`.

Closing as answered.
